### PR TITLE
Use PyLint 3.0+

### PR DESCRIPTION
**Pull request by sbrunner**
Based on #671, fix some errors

### CODE REVIEW

**Review by Pierre-Sassoulas — COMMENTED** [review]
I did not try to just vendor the missing deprecated function in prospector yet, did you think about it ?

**Comment by sbrunner** [comment]
When I saw the cleaning pull request, the missing function uses a lot of deprecated, then removed objects, then I didn't think that's easy to do!

File: CHANGELOG.rst [hunk 1]
```diff
@@ -2,6 +2,13 @@
 Changelog
 #########
 
+Version 1.11.1
+--------------
+
+**New**:
+
+* pylint version can now be above 3.0.0 again
```
Comment (Pierre-Sassoulas) [inline CHANGELOG.rst:10]: I used pylint > 3 in pyproject.toml for easy testing, if we want compatibility to pylint 2.8.3+ this is going to be harder. But imo it's reasonable to use only 3.+ It supports python 3.7 to 3.13

File: prospector/tools/pylint/linter.py [hunk 2]
```diff
@@ -12,13 +17,106 @@ def __init__(self, found_files, *args, **kwargs):
...
+        try:
+            self._parse_configuration_file(config_args)
+        except _UnrecognizedOptionError as exc:
+            unrecognized_options_message = ", ".join(exc.options)
+
+        # Then, if a custom reporter is provided as argument, it may be overridden
+        # by file parameters, so we re-set it here. We do this before command line
+        # parsing, so it's still overridable by command line options
+        # if reporter:
+        #     self.set_reporter(reporter)
```
Comment (sbrunner) [inline prospector/tools/pylint/linter.py:49]: It should be removed

File: prospector/tools/pylint/linter.py [hunk 3]
```diff
@@ -12,13 +17,106 @@ def __init__(self, found_files, *args, **kwargs):
...
+        # Now we parse any options from the command line, so they can override
+        # the configuration file
+        # args_list = _order_all_first(args_list, joined=True)
+        # parsed_args_list = self._parse_command_line_configuration(args_list)
+
+        # Remove the positional arguments separator from the list of arguments if it exists
+        # try:
+        #     parsed_args_list.remove("--")
+        # except ValueError:
+        #     pass
```
Comment (sbrunner) [inline prospector/tools/pylint/linter.py:64]: It should be removed

File: prospector/tools/pylint/linter.py [hunk 4]
```diff
@@ -12,13 +17,106 @@ def __init__(self, found_files, *args, **kwargs):
...
+        #     if opt.startswith("--"):
+        #         unrecognized_options.append(opt[2:])
+        #     elif opt.startswith("-"):
+        #         unrecognized_options.append(opt[1:])
+        if unrecognized_options:
+            msg = ", ".join(unrecognized_options)
+            try:
+                self._arg_parser.error(f"Unrecognized option found: {msg}")
+            except SystemExit:
+                sys.exit(32)
```
Comment (sbrunner) [inline prospector/tools/pylint/linter.py:78]: It should be removed

File: prospector/tools/pylint/linter.py [hunk 5]
```diff
@@ -12,13 +17,106 @@ def __init__(self, found_files, *args, **kwargs):
...
+                warnings.warn_explicit(
+                    f"'{exc_name}' is not a proper value for the 'overgeneral-exceptions' option. "
+                    f"Use fully qualified name (maybe 'builtins.{exc_name}' ?) instead. "
+                    "This will cease to be checked at runtime when the configuration "
+                    "upgrader is released.",
+                    category=UserWarning,
+                    filename="pylint: Command line or configuration file",
+                    lineno=1,
+                    module="pylint",
+                )
```
Comment (sbrunner) [inline prospector/tools/pylint/linter.py:98]: Shouldn't it be removed?

Comment (Pierre-Sassoulas) [reply prospector/tools/pylint/linter.py:98]: Definitely, this is a runtime check, user should have seen it between 2.16 and 3.0

File: CHANGELOG.rst [hunk 6]
```diff
@@ -2,6 +2,13 @@
 Changelog
 #########
 
+Version 1.11.1
```
Comment (sbrunner) [inline CHANGELOG.rst:5]: Shouldn't be an 1.12.0?

File: tests/tools/pylint/test_pylint_tool.py [hunk 7]
```diff
@@ -91,6 +91,8 @@ def test_will_throw_useless_suppression(self):
 
         found_files = _get_test_files("testpath", "testpath/test_useless_suppression.py")
         pylint_tool.configure(config, found_files)
+        # useless-suppression is now disable by default in pylint
```
Comment (sbrunner) [inline tests/tools/pylint/test_pylint_tool.py:94]: See: https://pylint.pycqa.org/en/latest/user_guide/messages/information/useless-suppression.html

File: CHANGELOG.rst (hunk 6 above)
Comment (Pierre-Sassoulas) [reply CHANGELOG.rst:5]: Yeah 👍 I was thinking it was only a simple formality of removing the <3.0 and testing that it worked originally :D (how naïve)

Comment (sbrunner) [reply CHANGELOG.rst]: Done

**Comment by sbrunner** [comment]
@Pierre-Sassoulas @carlio Do you think that's ready to be merged?

As I currently understand, if we want a better (more direct) implementation, we should do some modifications directly in PyLint!

**Review by Pierre-Sassoulas — APPROVED** [review]
Looks great !

**Review by carlio — APPROVED** [review]
Looks good to me

**Comment by sbrunner** [comment]
Thank you :-)

Do we want something other in version 1.12 or do we do the release?

**Comment by carlio** [comment]
I think it's best to release it now as people are waiting for a Python 3.12 compatible version; any other bugfixes or cleanups can go into a patch version.

I am pretty sure the PyPI token secrets in the repo are still correct so the GitHub action can still be used to make the release, if you want to.

**Comment by Pierre-Sassoulas** [comment]
You need to push a tag then create a github release from the tag to release on pypi @sbrunner (I could do it if you want, but it's a good thing if you already did it once imo)

**Comment by carlio** [comment]
https://github.com/prospector-dev/prospector/blob/master/pyproject.toml#L3 <-- need to update that :)

**Pull request merged**